force-app/main/default/classes/DMLOperations.cls

```diff
@@ -144,10 +186,66 @@ public with sharing class DMLOperations {
      * @param oppNames The list of Opportunity names.
      */
     public static void upsertOpportunities(String accountName, List<String> oppNames) {
-        // Loop through each Opportunity name
-            // Create an Opportunity with the given name
-            // Add the Opportunity to the list
+    // Query for the Account
+    List<Account> accList = [SELECT Id FROM Account WHERE Name = :accountName LIMIT 1];
+
+    Account acc;
+
+    if (!accList.isEmpty()) {
+        // Account with the given name exists, use it
+        acc = accList[0];
+    } else {
+        // Account with the given name doesn't exist, create a new one
+        acc = new Account(Name = accountName);
+        insert acc;
     }
+
+    // Create a list to store existing Opportunities
+    List<Opportunity> existingOpps = [SELECT Id, Name FROM Opportunity WHERE AccountId = :acc.Id];
+
+    // Create a map to store existing Opportunities by name
+    Map<String, Opportunity> oppMap = new Map<String, Opportunity>();
+    for (Opportunity opp : existingOpps) {
+        oppMap.put(opp.Name, opp);
+    }
+
+    // Create a list to store Opportunity records
+    List<Opportunity> opportunities = new List<Opportunity>();
+
+    // Loop through each Opportunity name
+    for (String oppName : oppNames) {
+        // Check if an Opportunity with the same name exists
+        if (oppMap.containsKey(oppName)) {
+            // If it exists, update it
+            Opportunity existingOpp = oppMap.get(oppName);
+            existingOpp.StageName = 'Prospecting';
+            existingOpp.CloseDate = Date.today().addDays(30);
+            opportunities.add(existingOpp);
+        } else {
+            // If it doesn't exist, create a new one
+            Opportunity newOpp = new Opportunity(
+                Name = oppName,
+                AccountId = acc.Id,
+                StageName ='Prospecting',
+                CloseDate = Date.today().addDays(30)
+            );
+            opportunities.add(newOpp);
+        }
+    }
+
+    // Insert or update the Account
+    if (acc.Id == null) {
+        insert acc;
+    } else {
+        update acc;
+    }
+
+    // Upsert the Opportunities
+    upsert opportunities;
```

Comment:
Good work. This is a hard problem. Not sure you need that if else condition from  line 237 to 241. You already inserted the account. And nothing needs updating. We just need the ID.